crates/worker/src/partition/snapshot_producer.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2023 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use anyhow::{anyhow, bail};
+use parking_lot::Mutex;
+use tokio::sync::oneshot::Sender;
+use tracing::{info, warn};
+
+use restate_core::TaskKind;
+use restate_partition_store::snapshots::{PartitionSnapshotMetadata, SnapshotFormatVersion};
+use restate_partition_store::PartitionStore;
+use restate_types::config::Configuration;
+use restate_types::identifiers::{PartitionId, SnapshotId};
+use restate_types::live::LiveLoad;
+use restate_types::time::MillisSinceEpoch;
+
+pub struct SnapshotProducer {
+    pub partition_id: PartitionId,
+    pub config: Box<dyn LiveLoad<Configuration> + Send>,
+    latest_snapshot: Arc<Mutex<Option<SnapshotId>>>,
+}
+
+impl SnapshotProducer {
+    pub fn new(partition_id: PartitionId, config: Box<dyn LiveLoad<Configuration> + Send>) -> Self {
+        Self {
+            partition_id,
+            config,
+            latest_snapshot: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    /// Creates a snapshot in a background task, optionally sending a completion notification.
+    pub(crate) fn create_snapshot(
+        &mut self,
+        partition_store: &mut PartitionStore,
+        response_tx: Option<Sender<anyhow::Result<SnapshotId>>>,
+    ) -> anyhow::Result<()> {
+        let partition_id = self.partition_id;
+
+        let arc = self.latest_snapshot.clone();
+        let lock = arc
+            .try_lock_arc()
+            .ok_or(anyhow!("Snapshot already in progress"))?;
```

Comment:
instead do 
```rust
use anyhow::Context;

// then to handle an error you can simply do
arc.try_lock_arc().context("snapshot already in progress")?;
```

Reply:
I tried the Tokio Mutex on Friday but I only realized now that I had to take an owned mutex guard to move it to the future! This is what I needed in the end:

```rs
let guard = self.latest_snapshot.clone()
    .try_lock_owned()
    .context(anyhow!("Snapshot already in progress"))?;
```